_posts/2024-01-22-Python-Memory-Management-and-Troubleshooting.md

```diff
@@ -0,0 +1,542 @@
+"Computer memory stores information, such as data and programs for immediate use in the computer. ... Main memory operates at a high speed compared to non-memory storage which is slower but less expensive and oftentimes higher in capacity. " ([Wikipedia: Computer memory](https://en.wikipedia.org/wiki/Computer_memory)).
+
+<table>
+<tr><th colspan="2">Memory Blocks</th></tr>
+<tr>
+<td>
+
+<strong>A.)</strong> All memory blocks available.
+
+<table>
+<tr><td>Block</td><td>Block</td><td>Block</td></tr>
+</table>
+
+</td>
+<td>
+
+<strong>B.)</strong> Some memory blocks in use.
+
+<table>
+<tr><td style="background:#86EFAC;">Block</td><td style="background:#86EFAC;">Block</td><td>Block</td></tr>
+</table>
+
+</td>
+</tr>
+<tr>
+<td colspan="2" style="text-align:center;font-weight:bold;">Practical analogy</td>
+</tr>
+<tr>
+<td>
+
+<strong>C.)</strong> You have limited buckets to hold things.
+
+<table>
+<tr><td>🪣</td><td>🪣</td><td>🪣</td></tr>
+</table>
+
+</td>
+<td>
+
+<strong>D.)</strong> Two buckets are used, the other remains empty.
+
+<table>
+<tr><td style="background:#86EFAC;">🪣</td><td style="background:#86EFAC;">🪣</td><td>🪣</td></tr>
+</table>
+
+</td>
+</tr>
+</table>
+
+_Fixed-size memory blocks may be free or used at various times. They can be thought of like reusable buckets to hold things._
+
+One way to organize computer memory is through the use of ["fixed-size blocks"](https://en.wikipedia.org/wiki/Memory_management#FIXED-SIZE), also called "blocks".
+Fixed-size memory blocks are chunks of memory of a certain byte size (usually all the same size).
+Memory blocks may be in use or free at different times.
+
+```mermaid!
+flowchart LR
+
+subgraph computer ["Computer (resources)"]
+
+subgraph memory["Memory"]
+    subgraph heap1 ["heap 1"]
+        direction TB
+        subgraph poola["pool a"]
+          blocks1["blocks"]
+        end
+        subgraph poolb["pool b"]
+          blocks2["blocks"]
+        end
+    end
+    subgraph heap2 ["heap 2"]
+        subgraph poolc["pool c"]
+          blocks3["blocks"]
+        end
+    end
+end
+
+end
+
+style computer fill:#fff,stroke:#333
+style memory fill:#86EFAC,stroke:#333
+style poola fill:#BFDBFE,stroke:#333;
+style poolb fill:#BFDBFE,stroke:#333;
+style poolc fill:#BFDBFE,stroke:#333;
+```
+
+_Memory heaps help organize available memory on a computer for specific procedures. Heaps may have one or many memory pools._
+
+Computer memory blocks may be organized in hierarchical layers to manage memory efficiently or towards a specific purpose.
+One top-level organization model for computer memory is through the use of ___heaps___ which help describe chunks of the total memory available on a computer for specific processes.
+These heaps may be ___private___ (only available to a specific software process) or ___shared___ (available to one or many software processes).
```

Comment:
Thanks for this. I tried finding more information and standard documentation covering what private means in this context and what the opposite or other options might be. Generally I found a gap from Python's documentation here where "private" is used but there's no definition of what private means for Python (https://docs.python.org/3/c-api/memory.html). Since it's used quite frequently I still label the Python heap as being private, but have removed the mention about shared memory as I couldn't seem to find a definition for this.